### ISSUE TITLE
Regression: Nested dictionaries fail in 3.4.0+ (after #3021)

### ISSUE THREAD

**Issue by Shane32** [linq2db/linq2db#3160]
This query fails for all builds after PR #3021 :

```cs
var ret = db.GetTable<Table1>()
	.Select(t1 => new Dictionary<string, object?>()
	{
		{
			"t1",
			db.GetTable<Table2>()
				.Where(x => x.ParentId == t1.Id)
				.Select(t2 => new Dictionary<string, object>()
				{
					{
						"t2",
						db.GetTable<Table3>()
							.Where(x => x.ParentId == t2.Id)
							.Select(t3 => t3.Id)
							.FirstOrDefault()
					}
				})
				.FirstOrDefault()
		}
	})
	.ToList();
```

The exception is:

```
The number of generic arguments provided doesn't equal the arity of the generic type definition. (Parameter 'instantiation')
```

The code works correctly in linq2db 3.3.0

### Environment details
linq2db version: 3.4.0 and newer including 4.0.0 beta
Database Server: SQL Server
Operating system: Windows 10
.NET Framework: .NET 5

**Comment by sdanyliv** [comment]
@Shane32, could you please post model used in this test?

**Comment by Shane32** [comment]
See PR #3162

**Comment by Shane32** [comment]
Branch issues/3160

**Comment by sdanyliv** [comment]
@Shane32, it is really worked before?

**Comment by sdanyliv** [comment]
I'll post workaround before I find time to resolve this issue:
```cs
var query = 
    from t1 in db.GetTable<Table1>()
    from t2 in db.GetTable<Table2>().Where(x => x.ParentId == t1.Id).Take(1).DefaultIfEmpty()
    from t3 in db.GetTable<Table3>().Where(x => x.ParentId == t2.Id).Take(1).DefaultIfEmpty()
    select new Dictionary<string, object?>()
    {
        {
            "t1",
            new Dictionary<string, object>()
                {
                    {
                        "t2",
                        t3.Id
                    }
                }
        }
    };

var ret = query.ToList();
```

**Comment by Shane32** [comment]
After #3028 merged, before #3021 was merged, the tests within #3162 run as follows:

![image](https://user-images.githubusercontent.com/6377684/128760954-8caa0c5f-7a1e-4e6d-b708-0050c552bfc7.png)

My specific use case is most similar to `NestedDictionaryTest3CrossApply`.

After #3021 was merged (the next commit in the history of `master`), the tests within #3162 run as follows:

![image](https://user-images.githubusercontent.com/6377684/128761263-4470ddee-d17c-480d-85cd-0e0184120f3e.png)

**Comment by Shane32** [comment]
Thanks for the workaround, but my queries are all built dynamically from metadata, which is why they use dictionaries in the first place.  The outer layer of the select has no knowledge of the inner layers' dependent tables, so the solution you proposed doesn't really work for me -- at least, not without a very large amount of effort.  I will simply stay on 3.3.0 until we can find a solution here.

**Comment by Shane32** [comment]
I've been working on these issues today.  I've found that the 'The number of generic arguments provided doesn't equal the arity of the generic type definition.` is due to a bug in this code here within EagerLoading.cs:

```cs
public static LambdaExpression CorrectLambdaType(LambdaExpression before, LambdaExpression after, MappingSchema mappingSchema)
{
	if (IsEnumerableType(before.ReturnType, mappingSchema) && before.ReturnType.IsGenericType
		/* patch as described in this post - see below */
		// && before.ReturnType.GetGenericTypeDefinition() == typeof(IEnumerable<>)
		)
	{
		var generic     = before.ReturnType.GetGenericTypeDefinition();
		var elementType = GetEnumerableElementType(after.ReturnType, mappingSchema);
		var desiredType = generic.MakeGenericType(elementType);
		if (after.ReturnType != desiredType && IsEnumerableType(after.ReturnType, mappingSchema))
		{
			after = Expression.Lambda(Expression.Convert(after.Body, desiredType), after.Parameters);
		}
	}

	return after;
}
```

https://github.com/linq2db/linq2db/blob/4d8cdc56c7bdc4d916569633eb22cf875a28e374/Source/LinqToDB/Linq/Builder/EagerLoading.cs#L1534..L1547

It's trying to detect the `IEnumerable` type of a dictionary, which is `KeyValuePair<x, y>` and then with `MakeGenericType` try to reconstruct the `Dictionary<,>` using `KeyValuePair<x, y>` as the type argument, which doesn't work.

I can reduce the scope of the condition by adding this phrase to the end of the first `if` condition:

```cs
 && before.ReturnType.GetGenericTypeDefinition() == typeof(IEnumerable<>)
```

Note that commenting out this entire method only causes a failure in `Issue1862TestProjections` test.  With the above patch, most, but not all, of the tests run successfully.

I've also discovered that the tests I wrote are executing as two commands, not a single one.  Prior to PR #3021, when these tests would work fine, they ran as a single execution.  These queries do not require eager loading as they can (and did) utilize OUTER APPLY to execute properly within a single query.

Here is a sample of the output of `NestedDictionaryTest2` (which works after the patch noted above within `EagerLoading.cs`):

```sql
DECLARE @take Int -- Int32
SET     @take = 1

SELECT
	[t1_1].[ID],
	[t1].[ID]
FROM
	[TABLE1] [t1_1]
		INNER JOIN [TABLE2] [detail] ON [detail].[PARENTID] = [t1_1].[ID]
		OUTER APPLY (
			SELECT TOP (@take)
				[_x].[ID]
			FROM
				[TABLE3] [_x]
			WHERE
				[_x].[PARENTID] = [detail].[ID]
		) [t1]
```

Followed by:

```sql
SELECT
	[t1].[ID]
FROM
	[TABLE1] [t1]
```

Whereas before PR #3021:

```sql
DECLARE @take Int -- Int32
SET     @take = 1
DECLARE @take_1 Int -- Int32
SET     @take_1 = 1

SELECT
	[t2].[ID]
FROM
	[TABLE1] [t1_1]
		OUTER APPLY (
			SELECT TOP (@take)
				[t1].[ID]
			FROM
				[TABLE2] [x_1]
					OUTER APPLY (
						SELECT TOP (@take_1)
							[x].[ID]
						FROM
							[TABLE3] [x]
						WHERE
							[x].[PARENTID] = [x_1].[ID]
					) [t1]
			WHERE
				[x_1].[PARENTID] = [t1_1].[ID]
		) [t2]
```

So it would seem that while the patch fixes a problem that can occur in rare circumstances, this is not the real problem here, and the `HasSubQuery` method added by PR #3021 is seriously flawed in its "more aggressive eager loading detection" logic.  I anticipate that it needs to be altered to only detect subqueries that return more than one record and hence _require_ eager loading.

**Comment by Shane32** [comment]
I'll continue investigation, but in the meantime, @sdanyliv I'd appreciate your comments on this - thanks!

**Comment by Shane32** [comment]
Solved be reverting #3021 via #3384
